### ISSUE TITLE
[Bug] - Memory Leaks

### ISSUE THREAD

**Issue by DerekCorniello** [DerekCorniello/mux-lang#196]
@jiraten / @ethantiller found a few memory leaks in the runtime (map, object, set, std, sync), so creating an issue for their branch #195 [DRAFT]. Looks like there are a few more. I used AI to find, so take it with a grain of salt:

# Memory leaks & memory-safety issues in `mux-runtime`

## Fixed in PR #195 (record only)

- `map.rs:43` `mux_map_value` — leaked `*mut Map` wrapper. Fixed.
- `set.rs:33` `mux_set_value` — leaked `*mut Set` wrapper. Fixed.
- `std.rs:99` `mux_list_value` — leaked `*mut List` wrapper. Fixed.
- `sync.rs:380` `mux_sync_spawn` — use-after-free on `*mut ClosureRepr` in
  spawned thread. Fixed.
- `object.rs` — added `copy` callback hook to `ObjectType` and
  `mux_register_object_copy`. API only; see issue 4 below.

### 1. `mux_tuple_value` leaks `*mut Tuple` wrapper

**Location:** `mux-runtime/src/tuple.rs:17-21`

**Snippet:**
```rust
pub extern "C" fn mux_tuple_value(tuple: *mut Tuple) -> *mut Value {
    let tuple = unsafe { &*tuple };
    let value = Value::Tuple(Box::new(tuple.clone()));
    mux_rc_alloc(value)
}
```

**Explanation:** Same bug the PR fixed for list/map/set, just missed for
tuples. `mux_new_tuple` (`tuple.rs:9`) does `Box::into_raw(Box::new(Tuple(...)))`
so the caller transfers ownership. The current code borrows + clones and
never frees the wrapper. Every tuple literal leaks.

**Suggested fix:**
```rust
let owned = unsafe { Box::from_raw(tuple) };
mux_rc_alloc(Value::Tuple(Box::new(owned)))
```

---

### 2. `in` operator on lists/sets leaks the extracted `*mut List/Set`

**Location:** `mux-compiler/src/codegen/operators.rs:805-820`

**Snippet:**
```rust
let raw_list = self.extract_list_from_value(right.into_pointer_value())?;
let item_ptr = self.ensure_pointer(left);
let result = self
    .generate_runtime_call("mux_list_contains", &[raw_list.into(), item_ptr.into()])
    .ok_or("mux_list_contains returned no value")?;
Ok(result)
```

**Explanation:** `mux_value_get_list` / `mux_value_get_set` (`std.rs:112, 140`)
return a heap-allocated `*mut List/Set` the caller must free.
`mux_list_contains` / `mux_set_contains` only borrow it. The raw pointer is
never freed. Leaks on every list-membership and set-membership check.

**Suggested fix:** Call `mux_free_list` / `mux_free_set` on the raw pointer
after the contains check (or, better, change `mux_list_contains` /
`mux_set_contains` to take `*const Value` and read the inner collection).

---

### 3. `copy_object` API added but never applied to existing non-POD types

**Location:** `mux-runtime/src/sync.rs:284-294`, `sql.rs:35-49`,
`net.rs:26-41`

**Explanation:** `Mutex`, `RwLock`, `CondVar`, `Thread`, `SqlConnection`,
`SqlTransaction`, `SqlResultSet`, `TcpStream`, `TcpListener`, `UdpSocket` all
register via the no-copy `register_object_type` and store a raw pointer in 8
bytes. If any is ever passed through `mux_copy_object` (the codegen does in
`expressions.rs:1316` and `:2969`), the byte-copy fallback copies the raw
pointer and both copies' destructors free the same pointer — double free. PR
#195 added the API to fix this but didn't migrate any of the affected types.

**Suggested fix:** Either register a copy callback for each (Thread could
share the `JoinHandle` via `Arc` + refcount; sql/net could return
not-copyable), or have `copy_object` return null for types without a copy
callback and have the codegen surface a "cannot copy" error.

---

### 4. `mux_thread_detach` drops `JoinHandle` without calling `detach()`

**Location:** `mux-runtime/src/sync.rs:461-479` and
`sync.rs:355-364` (`destroy_thread_object`)

**Snippet:**
```rust
let Some(entry) = threads.remove(&id) else { ... };
if entry.handle.is_none() {
    return err_string("Thread already joined or detached");
}
ok_unit()
// entry drops here, dropping Option<JoinHandle<()>>
```

**Explanation:** Dropping a `JoinHandle` without `join()` or `detach()` relies
on Rust's auto-detach, which is not guaranteed on all platforms and can panic.
Function is named "detach" but never explicitly calls `detach()`.
`destroy_thread_object` has the same issue when a `Thread` value is GC'd
without being joined.

**Suggested fix:**
```rust
if let Some(h) = entry.handle {
    h.detach();
}
```

---

### 5. FFI panics are undefined behavior

**Locations:**
- `mux-runtime/src/std.rs:209` (`mux_value_to_list`)
- `mux-runtime/src/int.rs:71` (`mux_int_from_value`)
- `mux-runtime/src/float.rs:82, 94, 105`
- `mux-runtime/src/bool.rs:58, 71`
- `mux-runtime/src/sync.rs:510, 531, 554` (`mux_mutex_new` /
  `mux_rwlock_new` / `mux_condvar_new`)
- `mux-runtime/src/object.rs:97` (`alloc_object`)
- `mux-runtime/src/refcount.rs:84` (`mux_rc_alloc`)

**Explanation:** Each of these `#[unsafe(no_mangle)] pub extern "C" fn` calls
`panic!(...)` on bad input. Panicking across the FFI boundary is undefined
behavior; the process aborts with no useful Mux-level error.

**Suggested fix:** Return `std::ptr::null_mut()` and have the codegen check
for null, or return a `Value::Result(Err(...))` so the error surfaces as a
Mux `Err` value.

---

### 6. Inconsistent ownership model in sync vs sql

**Locations:** `mux-runtime/src/sync.rs:508, 529, 552` vs
`mux-runtime/src/sql.rs:928`

**Snippet (sync.rs):**
```rust
Box::into_raw(Box::new(value))   // raw Box pointer, no RefHeader
```

**Snippet (sql.rs):**
```rust
mux_rc_alloc(value)              // refcounted *mut Value, with RefHeader
```

**Explanation:** `mux_mutex_new` / `mux_rwlock_new` / `mux_condvar_new` return
a raw `Box` pointer; `mux_sql_connect` returns a refcounted pointer. These
are different memory layouts. If the codegen ever calls `mux_rc_inc` /
`mux_rc_dec` on the sync.rs values, it reads garbage from the supposed
header location and corrupts memory.

**Suggested fix:** Decide on one model. If refcounted, replace
`Box::into_raw(Box::new(value))` with `mux_rc_alloc(value)` in the three
sync functions.
